lib/elixir/lib/code/normalizer.ex

```diff
@@ -0,0 +1,456 @@
+defmodule Code.Normalizer do
+  @moduledoc false
+
+  defguard is_literal(x)
+           when is_integer(x) or
+                  is_float(x) or
+                  is_binary(x) or
+                  is_atom(x)
+
+  @doc """
+  Wraps literals in the quoted expression to conform to the AST format expected
+  by the formatter.
+  """
+  def normalize(quoted, opts \\ []) do
+    line = Keyword.get(opts, :line, 1)
+    escape = Keyword.get(opts, :escape, true)
+
+    state = %{
+      escape: escape,
+      parent_meta: [line: line],
+      escape_interpolation: false
+    }
+
+    do_normalize(quoted, state)
+  end
+
+  # Skip normalized literals
+  defp do_normalize({:__block__, _, [literal]} = quoted, _state) when is_literal(literal) do
+    quoted
+  end
+
+  # Skip normalized charlists
+  defp do_normalize({:__block__, meta, [charlist]} = quoted, state)
+       when is_list(charlist) do
+    if Keyword.has_key?(meta, :delimiter) do
+      quoted
+    else
+      do_normalize(charlist, state)
+    end
+  end
+
+  # Only normalize the first argument of an alias if it's not an atom
+  defp do_normalize({:__aliases, meta, [first | rest]}, state) when not is_atom(first) do
+    meta = patch_meta_line(meta, state.parent_meta)
+    first = do_normalize(first, %{state | paret_meta: meta})
+    {:__aliases__, meta, [first | rest]}
+  end
+
+  defp do_normalize({:__aliases__, _, _} = quoted, _state) do
+    quoted
+  end
+
+  # Skip captured arguments like &1
+  defp do_normalize({:&, meta, [term]}, state) when is_integer(term) do
+    meta = patch_meta_line(meta, state.parent_meta)
+    {:&, meta, [term]}
+  end
+
+  # Ranges
+  defp do_normalize(left..right//step, state) do
+    left = do_normalize(left, state)
+    right = do_normalize(right, state)
+
+    meta = [line: state.parent_meta[:line]]
+
+    if step == 1 do
+      {:.., meta, [left, right]}
+    else
+      step = do_normalize(step, state)
+      {:"..//", meta, [left, right, step]}
+    end
+  end
+
+  # Bit containers
+  defp do_normalize({:<<>>, meta, parts} = quoted, state) do
+    meta = patch_meta_line(meta, state.parent_meta)
+
+    parts =
+      if interpolated?(quoted) do
+        normalize_interpolation_parts(parts, %{state | parent_meta: meta})
+      else
+        Enum.map(parts, &do_normalize(&1, %{state | parent_meta: meta}))
+      end
+
+    {:<<>>, meta, parts}
+  end
+
+  # Atoms with interpolations
+  defp do_normalize(
+         {{:., dot_meta, [:erlang, :binary_to_atom]}, call_meta, [{:<<>>, _, _} = string, :utf8]},
+         state
+       ) do
+    dot_meta = patch_meta_line(dot_meta, state.parent_meta)
+    call_meta = patch_meta_line(call_meta, dot_meta)
+
+    string =
+      if state.escape do
+        do_normalize(string, %{state | escape_interpolation: true})
```

Comment:
You need to be careful with this kind of flags because you can have this:

    :"foo#{~s/\n/}bar"

And escaping the interpolation inside the atom will now force the interpolation to be escaped inside the sigil. It is probably best to, once you know you have a `{:<<>>, ..., ...}`, you call a explicit `do_normalize_interpolation` with the escape as argument.

Reply:
I moved the bitstring normalization code into its own function with an explicit argument to avoid that issue